File: ast_printer.c
#include "ast_printer.h"
#include "ast.h"
#include "defs.h"
#include "error.h"
#include <stdio.h>

void print_var(size_t value, ASTMode mod) {
  switch (mod) {
  case VarName:
    printf("%s", (char *)value);
    break;
  case StackLoc:
    printf("(deref rbp %ld)", value * -8);
    break;
  default:
    error("unexpected value");
  }
}

void print_ast(ASTNode *node, ASTMode mod) {
  switch (node->token) {
  case Fixnum:
    printf("%ld", node->value);
    break;
  case Neg:
    printf("(- ");
    print_ast(node->lhs, mod);
    printf(")");
    break;
  case Add:
    printf("(+ ");
    print_ast(node->lhs, mod);
    printf(" ");
    print_ast(node->rhs, mod);
    printf(")");
    break;
  case Read:
    printf("(read)");
    break;
  case Var:
    print_var(node->value, mod);
    break;
  case Let:
    printf("(let ([%s %ld]) ", (char *)node->lhs->value, node->value);
    print_ast(node->rhs, mod);
    printf(")");
    break;
  case REG:
    printf("(reg ");
    switch (node->value) {
    case RAX:
      printf("RAX");
      break;
    default:
      error("unexpected reg");
    }
    printf(")");
    break;
  case Assign:
    printf("(assign ");
    print_var(node->value, mod);
    printf(" ");
    print_ast(node->lhs, mod);
    printf(")");
    break;
  case MOVQ:
    printf("MOVQ ");
    print_ast(node->lhs, mod);
    printf(" ");
    print_var(node->value, mod);
    break;
  case ADDQ:
    printf("ADDQ ");
    print_ast(node->lhs, mod);
    printf(" ");
    print_var(node->value, mod);
    break;
  case CALLQ:
    printf("CALLQ %s", (char *)node->value);
    break;
  default:
    printf("\nprint_ast: failed to parse token %d\n", node->token);
    exit(-1);
  }
}

File: src/ast_printer.rs
use crate::ast::{Node, Token, Value};

pub fn print_ast(node: Box<Node>) {
    use Token::*;
    match node.token {
        Fixnum => match node.value.expect("value") {
            Value::Fixnum(num) => {
                print!("{}", num);
            }
            val => {
                panic!("unexpected {:?}", val);
            }
        },
        Program => {
            print!("(program ");
            print_ast(node.lhs.expect("lhs"));
            print!(")");
        }
        Neg => {
            print!("(- ");
            print_ast(node.lhs.expect("lhs"));
            print!(")");
        }
        Add => {
            print!("(+ ");
            print_ast(node.lhs.expect("lhs"));
            print!(" ");
            print_ast(node.rhs.expect("rhs"));
            print!(")");
        }
        Read => {
            print!("(read)");
        }
        Var => match node.value.expect("value") {
            Value::Var(var) => print!("{}", var),
            val => {
                panic!("unexpected {:?}", val);
            }
        },
        STACK_LOC => {
            print!("(deref RBP {:?})", node.value);
        }
        Let => match node.value.expect("value") {
            Value::Let(var, num) => {
                print!("(let ([{} {}]) ", var, num);
                print_ast(node.lhs.expect("lhs"));
                print!(")");
            }
            val => panic!("unexpected {:?}", val),
        },
        //   REG => {
        //     print!("(reg ");
        //     match (node.value) {
        //     RAX =>
        //       print!("RAX"),
        //     _ =>
        //       panic!("unexpected reg"),
        //     }
        //     print!(")");
        //   }
        //   Assign => {
        //     print!("(assign ");
        //     print_ast((ASTNode *)node.value);
        //     print!(" ");
        //     print_ast(node.lhs);
        //     print!(")");
        //   }
        //   MOVQ => {
        //     print!("MOVQ ");
        //     print_ast(node.lhs);
        //     print!(" ");
        //     print_ast((ASTNode *)node.value);
        //   }
        //   ADDQ => {
        //     print!("ADDQ ");
        //     print_ast(node.lhs);
        //     print!(" ");
        //     print_ast((ASTNode *)node.value);
        //   }
        //   CALLQ => {
        //     print!("CALLQ %s", (char *)node.value);
        //   }
        _ => {
            panic!("\nprint_ast: failed to parse token {:?}", node.token);
        }
    }
}

// void print_stmt(ASTNode *p) {
//   while (p != 0) {
//     print_ast(p);
//     print!("\n");
//     p = p.rhs;
//   }
// }
